crates/holochain_sqlite/tests/db_wrapper.rs

```diff
@@ -0,0 +1,383 @@
+        let my_handle = db_handle.clone();
+        let my_number_of_connections_received = number_of_connections_received.clone();
+        join_handles.push(std::thread::spawn(move || {
+            // Number of database connections opened over time
+            for _ in 0..100 {
+                let _ = my_handle.conn().unwrap();
+                my_number_of_connections_received.fetch_add(1, Ordering::SeqCst);
+
+                std::thread::sleep(std::time::Duration::from_millis(100));
+            }
+        }));
+    }
+
+    for jh in join_handles.into_iter() {
+        jh.join().unwrap();
+    }
+
+    assert_eq!(
+        20_000,
+        number_of_connections_received.load(Ordering::Acquire)
+    )
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[test]
+fn pool_size_is_limited() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    // Assumes knowledge of the implementation, readers plus one writer
+    let pool_size = num_read_threads() + 1;
+
+    let mut hold_connections = Vec::with_capacity(pool_size);
+    for _ in 0..pool_size {
+        let c = db_handle.conn().unwrap();
+        hold_connections.push(c)
+    }
+
+    let mut failed_count = 0;
+    for _ in 0..3 {
+        // Cannot get any more connections on this thread
+        match db_handle.conn() {
+            // This is caused by a timeout at the connection pool level but could be more specific
+            Err(DbConnectionPoolError(_)) => {
+                // Could not get a connection, this is the desired outcome. Sleep and try again
+                failed_count += 1;
+                std::thread::sleep(std::time::Duration::from_millis(10));
+            }
+            Err(e) => {
+                panic!("Got an unexpected error - {:?}", e);
+            }
+            Ok(_c) => {
+                panic!("Should not have been able to get a connection");
+            }
+        }
+    }
+
+    assert_eq!(3, failed_count);
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn reader_permits_are_limited() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let num_readers = num_read_threads();
+
+    let mut held_permits = Vec::with_capacity(num_readers);
+    for _ in 0..num_readers {
+        let c = db_handle.conn_permit::<DatabaseError>().await.unwrap();
+        held_permits.push(c)
+    }
+
+    let mut failed_count = 0;
+    for _ in 0..3 {
+        // Should not be able to get another permit
+        match db_handle.conn_permit().await {
+            Err(Timeout(_)) => {
+                // Could not get a permit, this is the desired outcome. Sleep and try again
+                failed_count += 1;
```

Comment:
It looks like the timeout for this is 10 seconds, making this a 30 second test. Should we make it possible to adjust this timeout for testing?

Reply:
Or, I *do* see these are marked slow_test... maybe we're just accepting that for now?

Reply:
We always run with slow tests turned on in CI because we have so many 'slow tests' that do take over 100ms or so. That makes it kind of meaningless to mark these slow tests as it only really helps people running locally for now.

I've tweaked the code to make the 'get a permit' and 'get a connection' timeouts configurable. I can't turn them down as low as I'd like because you get tests failing in the database initialisation code which is actually useful information that I didn't have before. I have made those times 100ms and 300ms respectively down from 10s and 30s. This means the tests finish in <2.5s rather than 90s. So quite a big difference now, plus some tests should be removed soon after they're put in as I've said above with removing code